runtime/bindings/python/hal.cc

```diff
@@ -1283,11 +1287,11 @@ void SetupHalBindings(nanobind::module_ m) {
 
   m.def(
```

Comment:
I think this API has exceeded its carrying capacity. Let's split it into three:

* get_cached_hal_driver()
* create_hal_driver()
* clear_hal_driver_cache()

You could merge the first and second by taking a bool `use_cache = True`. I slightly prefer the three-API form, though because "create" and "get_cached" mean different things.

Reply:
If we have separate `get_cached_hal_driver()` and `create_hal_driver()` functions, are users going to think that they should be doing all their driver creation with `create_hal_driver()` and only call `get_cached_hal_driver()` when they want to fetch one that they know they've already created?  That's what it would mean to me.  But then I'm weird.

Reply:
These are still "internal" APIs in that you get at them via the `iree.runtime._binding.get_cached_hal_driver()` namespace. So they are used as part of the user level APIs or for advanced/nefarious ends. 

Being explicit with the names is just me wanting glancibility when looking at implementations so it is clear what is using shared, global state and what is not.

Reply:
@stellaraccident What do you think about `get_hal_driver(uri, use_cache=True)`?  Maybe the concept of creation isn't expressed as strongly as in `create_hal_driver`, but "get" can represent both creation and getting a pre-existing object.  That would avoid the confusion I mentioned above.

Reply:
I guess my problem with `get_cached_hal_driver` is that it doesn't imply creating anything to me, much less so than `get_hal_driver()`.  I would think that that's a worse problem than not knowing what's using global state.  I don't think my API is ideal either, so if I've failed to convince you, you can make an executive decision and tell me what you'd like, and I'd be fine with it. :-)

While I'm whining, "parse_flags" is a terrible name for this function.  It confused 100% of the developers who were looking at it with me.  "add_flags" would have been perfect.

Reply:
Oh, wasn't thinking about the fact that `get_cached_hal_driver` is already published, so there's nothing that can be done about that unfortunate name.  In that case, I fully agree with you, and I will make three separate functions.

Reply:
Done.